hudi-common/src/main/java/org/apache/hudi/avro/AvroCastingGenericRecord.java

```diff
@@ -0,0 +1,259 @@
+        };
+      }
+      case ENUM: {
+        ValidationUtils.checkArgument(
+            oldSchema.getType() == STRING || oldSchema.getType() == Schema.Type.ENUM,
+            "Only ENUM or STRING type can be converted ENUM type");
+        if (oldSchema.getType() == STRING) {
+          return s -> new GenericData.EnumSymbol(newSchema, s);
+        }
+        return null;
+      }
+      case ARRAY: {
+        UnaryOperator<Object> op = getConverter(oldSchema.getElementType(), newSchema.getElementType());
+        if (op == null) {
+          return null;
+        }
+        return a -> {
+          ValidationUtils.checkArgument(a instanceof Collection, "cannot rewrite record with different type");
+          Collection array = (Collection) a;
+          List<Object> newArray = new ArrayList(array.size());
+          for (Object element : array) {
+            newArray.add(op.apply(element));
+          }
+          return newArray;
+        };
+      }
+
+      case MAP: {
+        UnaryOperator<Object> op = getConverter(oldSchema.getValueType(), newSchema.getValueType());
+        if (op == null) {
+          return null;
+        }
+        return m -> {
+          ValidationUtils.checkArgument(m instanceof Map, "cannot rewrite record with different type");
+          Map<Object, Object> map = (Map<Object, Object>) m;
+          Map<Object, Object> newMap = new HashMap<>(map.size());
+          for (Map.Entry<Object, Object> entry : map.entrySet()) {
+            newMap.put(entry.getKey(), op.apply(entry.getValue()));
+          }
+          return newMap;
+        };
+      }
+      case UNION:
+        return getConverter(getActualSchemaFromUnion(oldSchema), getActualSchemaFromUnion(newSchema));
+      case STRING:
+        switch (oldSchema.getType()) {
+          case ENUM:
+            return String::valueOf;
+          case BYTES:
+            return b -> String.valueOf(((ByteBuffer) b));
+          case INT:
+          case LONG:
+          case FLOAT:
+          case DOUBLE:
+            return Object::toString;
+          default:
+            return null;
+        }
+      case BYTES:
+        if (oldSchema.getType() == STRING) {
+          return s -> ByteBuffer.wrap(getUTF8Bytes(s.toString()));
+        }
+        return null;
+      default:
+        return null;
+    }
+  }
+
+  private static Schema getActualSchemaFromUnion(Schema schema) {
+    Schema actualSchema;
+    if (!schema.getType().equals(UNION)) {
+      return schema;
+    }
+    if (schema.getTypes().size() == 2
+        && schema.getTypes().get(0).getType() == Schema.Type.NULL) {
+      actualSchema = schema.getTypes().get(1);
+    } else if (schema.getTypes().size() == 2
+        && schema.getTypes().get(1).getType() == Schema.Type.NULL) {
+      actualSchema = schema.getTypes().get(0);
+    } else if (schema.getTypes().size() == 1) {
+      actualSchema = schema.getTypes().get(0);
+    } else {
+      return schema;
```

Comment:
In other portions of the codebase, we treat anything with more than 2 types in the union as invalid and throw an exception. Should we do that here as well?

Reply:
The MDT uses a trick where they have way more than 2 types. See HoodieAvroUtils.wrapValueIntoAvro